### PR TITLE
Scale large texture to the actual display size

### PR DESCRIPTION
**Pull request by mxalbert1996**
## Description

After this PR, `CGUILargeTextureManager` will load the scaled version of the texture with the actual display size. The scaled version will be kept in memory until released. When a new size of a texture is requested, the texture will be loaded and scaled again. When the window size changes, the textures will be reloaded with their new sizes.

## Motivation and context

To address the long-standing GUI image quality problem: https://forum.kodi.tv/showthread.php?tid=200401
This greatly improves the display quality of artwork.

## How has this been tested?

Tested on a MacBook Pro (M1 Pro, macOS 14.5), a 3rd gen Fire TV Cube (Fire OS 7/Android 9) and a Chromecast with Google TV (4K) (Android 12).

### Performance

On all the devices I've tested on, this didn't cause any noticeable delay on image loading.
This might need to be tested on some lower end devices, though.

### Memory usage

This saves a lot of memory both on macOS (checked the memory usage of Kodi with Activity Monitor) and on Android TV (checked the memory usage of Kodi with `adb shell dumpsys meminfo`).

## Screenshots (if appropriate):

Before:
screenshot00000

After:
screenshot00001

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [x] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/guilib/Texture.h
Comment (mxalbert1996) [inline xbmc/guilib/Texture.h:52]: The `requirePixels` parameter was originally added in #3704 for RPi platform, but now RPi platform is removed and the parameter doesn't seem to be used anywhere, so it should be okay to remove it.

**Comment by sarbes** [comment]
I guess you are following the downscaling/mipmapping discussion in the forum and on Github. 

In general, this PR is going into the right direction. The results are how it should look like. But the implementation is not there yet.

Messing around with the picture handling (as in photo viewer) for scaling is not the right place. Use the existing scaling capability of the image loader pipeline: https://github.com/xbmc/xbmc/blob/e2723d6eac495035d988651874eaf7df7cddff4a/xbmc/guilib/Texture.cpp#L268
Currently, the scaling itself is done inside of our default image loader (and probably in every image loader add-on), but I'm planning on moving that "outside", so that all image loaders can profit: https://github.com/xbmc/xbmc/blob/e2723d6eac495035d988651874eaf7df7cddff4a/xbmc/guilib/FFmpegImage.cpp#L434

This would speed up the process of loading a the texture by a lot, by just having one swscale step. I would wager that you could drop almost half the code by just doing that.

On contrast, the performance of your implementation is lacking because the texture is converted twice, and is being memcopied afterwards.

**Comment by bossanova808** [comment]
Absolutely great to see some work in this area.

For me particularly, whilst I share the general poster/artwork concerns, I am most looking forward to the Photo Viewer also improving...at the moment, I maintain an entire second library of images just for Kodi use, all pre-scaled to 1920x1080, as if I e.g. feed Kodi 4k images (3840x2160) and let Kodi scale, the results are very significantly less pleasant than using the pre-scaled...whereas this is not true in just about any other image viewer (where the down-sampled 4k is as good or better than the Lightroom produced pre-scaled version).

On this note, I would recommend you consider looking at some different scaling algorithms than straight bicubic for downsizing.  Obviously a primary goal would be to not to have the jaggies/artefacts in the 'Before' image.  But _ideally_ there would be significantly higher detail retention as well - you can see the resulting softness of straight bicubic in the 'After' screenshot above.

In this world of 4K+ TVs sharper results will look better to most people.  E.g. 'bicubic sharper' tends to retain detail better, at a minimum.  But there are many other algorithms that are better again - although it seems maybe FFMPEG only offers a few of these (? https://ffmpeg.org/ffmpeg-scaler.html).  Not sure if that's the only scaling lib in Kodi?

Ideally for people who care, it would be configurable (so users can choose a nice balance of power vs, performance based on their particular hardware) - as a practical example, I use a 'cat-rom' implementation on a webserver, which is far better at detail retention again, and yet only marginally slower (on a VPS, and the results are then cached - of course testing would be required on many and various Kodi devices out there).  

So - a sensible, sharper default, plus an advanced setting to choose a preferred algorithm...would be ideal (and this working consistently across the Kodi UI and the Picture Viewer).

Apologies if this is the wrong place/time/approach.

**Comment by sarbes** [comment]
Please, do not focus on the scaling algorithm in this PR. Image minification has a lot of subtle nuances. For example, scaling with linear gamma has far more influence on image quality. But all this can be explored in a different PR.

**Comment by fritsch** [comment]
> Absolutely great to see some work in this area.
> 
> For me particularly, whilst I share the general poster/artwork concerns, I am most looking forward to the Photo Viewer also improving...at the moment, I maintain an entire second library of images just for Kodi use, all pre-scaled to 1920x1080, as if I e.g. feed Kodi 4k images (3840x2160) and let Kodi scale, the results are very significantly less pleasant than using the pre-scaled...whereas this is not true in just about any other image viewer (where the down-sampled 4k is as good or better than the Lightroom produced pre-scaled version).
> 
> On this note, I would recommend you consider looking at some different scaling algorithms than straight bicubic for downsizing. Obviously a primary goal would be to not to have the jaggies/artefacts in the 'Before' image. But _ideally_ there would be significantly higher detail retention as well - you can see the resulting softness of straight bicubic in the 'After' screenshot above.
> 
> In this world of 4K+ TVs sharper results will look better to most people. E.g. 'bicubic sharper' tends to retain detail better, at a minimum. But there are many other algorithms that are better again - although it seems maybe FFMPEG only offers a few of these (? https://ffmpeg.org/ffmpeg-scaler.html). Not sure if that's the only scaling lib in Kodi?
> 
> Ideally for people who care, it would be configurable (so users can choose a nice balance of power vs, performance based on their particular hardware) - as a practical example, I use a 'cat-rom' implementation on a webserver, which is far better at detail retention again, and yet only marginally slower (on a VPS, and the results are then cached - of course testing would be required on many and various Kodi devices out there).
> 
> So - a sensible, sharper default, plus an advanced setting to choose a preferred algorithm...would be ideal (and this working consistently across the Kodi UI and the Picture Viewer).
> 
> Apologies if this is the wrong place/time/approach.

https://kodi.wiki/view/Advancedsettings.xml#imagescalingalgorithm

**Comment by mxalbert1996** [comment]
Right, turns out it's just the correct size wasn't passed to `Decode` function. Using that did save a lot of code.
I've also added the ability to refresh large texture when the display size changes, so no need to restart after toggling the setting or resizing the window now.

Besides, I used `adb shell dumpsys meminfo` to check the memory usage of Kodi on Android TV and confirmed that enabling the setting reduces the memory usage dramatically (over 50% for me).

The commit to remove `requirePixels` parameter is not necessary. I kept it, but please let me know if you think it's better to drop it.

**Comment by sarbes** [comment]
> Right, turns out it's just the correct size wasn't passed to `Decode` function. Using that did save a lot of code. And the resulting image is actually slightly sharper. 

Excellent. It also avoids the need to mess with the texture type in the loader thread. You've assumed it to be BGRA, but it could also be DXT1/3/5, which would likely crash Kodi. There are actually a lot of texture formats (https://github.com/xbmc/xbmc/blob/master/xbmc/guilib/TextureFormats.h). The compressed ones can't even be scaled, but by passing the requested dimensions to the decoder, a suitable mip could be selected from a file which supports this (DDS, PKM, ASTC, ...).

> I've also added the ability to refresh large texture when the display size changes, so no need to restart after toggling the setting or resizing the window now.

Yes, this is important. The resolution might change automatically if the resolution whitelist for videos is enabled.

> Now there's no noticeable delay on all the devices I tested on. Besides, I used `adb shell dumpsys meminfo` to check the memory usage of Kodi on Android TV again and confirmed that enabling the setting actually reduces the memory usage dramatically (over 50% for me).

Good! This is obviously dependent on a lot of factors. What's even more important than the actual size in RAM is the saving in memory bandwidth. This might enable us to render at 60fps instead of 50fps, depending on the scenario.

> The commit to remove `requirePixels` parameter is not necessary. I kept it, but please let me know if you think it's better to drop it.

Just drop it if you like.

> Also, enabling the setting doesn't seem to cause any negative effect, so perhaps we can consider making it enabled by default, or even always enabled?

I don't see any reason to have a setting for it. I do gate some of my rendering improvements behind advanced settings, as some are a bit more complex. But this is straight forward. Perhaps we might run into trouble with elements with some funky zoom animations, but I wouldn't sweat it. Just remove the setting.

I do hope to have some time to do a proper review in the next few days. But other than 1-2 details, I'm pretty satisfied from a first glance.

**Review by sarbes — DISMISSED** [review]
Those are the blockers I see.

**Comment by sarbes** [comment]
The textures are close in size, but not exactly at rendering resolution. Because of this, there is an extra bilinear scaling step on the GPU, resulting in a softer image. For example, the posters in this scene render at 172x245, but the texture is 176x251 (same as the image, which is a subset of the texture):
image

The texture should have a resolution of 176x245. The image itself would be 172x245, with four columns of padding (clamped color).

File: xbmc/guilib/Texture.cpp
Comment (mxalbert1996) [reply xbmc/guilib/Texture.cpp:248]: > In the context of textures, this would be the imageWidth. It describes a subset of the textureWidth (which might be up to 15px wider). You need to pass this imageWidth to the decoder as second parameter.

Understood.

But is it right to ignore the original aspect ratio? If we do that, the aspect ratio of the `CGUITexture` (which comes from [the skin](https://kodi.wiki/view/Aspect_Ratio)) will not work and I don't think that's what we want.
https://github.com/xbmc/xbmc/blob/4e4b2888b41d9414a270a69d38dc96c356d29fe4/xbmc/guilib/GUITexture.h#L228

Comment (sarbes) [reply xbmc/guilib/Texture.cpp:248]: The skinning engine has multiple modes. It can keep the aspect of the texture, but it can also dictate arbitrary resolutions. I'll have to check what the best approach is, but I'm mobile ATM.

Comment (sarbes) [reply xbmc/guilib/Texture.cpp:248]: Unfortunately, we need to go the extra mile and propagate the aspect ratio "mode" (https://kodi.wiki/view/Aspect_Ratio) through. Based on this, the actual texture resolution has to be calculated.

Right now, the current implementation reduces the texture resolution to fit the "keep" aspect ratio. "scale" and "stretch" are often taking up more screen space, but have to use the overly minified texture. "center" is totally broken, as it is scaled to "keep" sizes.

**Comment by sarbes** [comment]
> Btw, I tried RenderDoc and can see the texture sizes. But how can I check at what size a texture is rendered?

You can measure it with an image editor (like Gimp). The mesh size should also correspondent to the element size.

I hope to review of the current state in the next few days.

File: xbmc/guilib/GUITexture.cpp
Comment (mxalbert1996) [inline xbmc/guilib/GUITexture.cpp:373]: One concern here is that, when being rendered, the size is the difference of the two positions that are rounded individually, so the size calculated here can differ by 1 pixel from the actual size.
I've thought of calculating the size like `(int)((m_posX + m_width) / guiScaleX) - (int)(m_posX / guiScaleX)`, but `m_posX` and `m_posY` isn't global anyway, so I'm not sure if that can really solve the problem.

Comment (sarbes) [reply xbmc/guilib/GUITexture.cpp:373]: I'd have to look at the element you have issues with, but I can think of scenarios where this might be the case.

Our skinning system operates on a fixed grid, defined by the skin. If the grid size differs to the rendering resolution (let's say, 1280x720 grid size to 1920x1080 screen), our skinning system "snaps" vertex coordinates to their next position on the screen grid. It is not unusual to see such an element having different dimensions depending on its position. I guess this is baggage from way back, where most of the graphics pipeline was fixed function/software driven.

I've started to modernize the font geometry handling (on GL/GLES), which had similar issues (https://github.com/xbmc/xbmc/pull/24605). Somewhere I do have a branch doing the same for textures, but it is not ready yet.

If I were you, I would only test 1080p skins at 1920x1080 or 4k, and ignore other resolutions for now.

Comment (mxalbert1996) [reply xbmc/guilib/GUITexture.cpp:373]: I've tested this at 4K and it's working as expected.

File: xbmc/guilib/FFmpegImage.cpp
Comment (neo1973) [inline xbmc/guilib/FFmpegImage.cpp:490]:
```suggestion
  SwsContext* context =
```

File: xbmc/guilib/AspectRatio.h
Comment (neo1973) [inline xbmc/guilib/AspectRatio.h:36]: I would prefer for this to be a `enum class` but it's not a blocker as it's only moved old code.
```suggestion
  enum class AspectRatio
  {
    // Do not keep aspect ratio. Image size = box size.
    STRETCH = 0,
    // Keep aspect ratio. Image size >= box size.
    SCALE,
    // Keep aspect ratio. Image size <= box size.
    KEEP,
    // Do not scale image.
    CENTER,
  };
  using enum AspectRatio;
```

File: xbmc/guilib/AspectRatio.h
Comment (neo1973) [inline xbmc/guilib/AspectRatio.h:21]: Lets make all theses `static constexpr unsigned int`.

**Review by neo1973 — COMMENTED** [review]
Looks promising :slightly_smiling_face:

I tested this on a RPi3B and didn't noticed delays loading textures. Rendering was probably a little faster depending on scene but I can't say for certain. It definitely saves a little of memory.

I wish the scaled textures were a little sharper but maybe this can be improved in the future using different scaling algorithms.

Code wise its pretty good, I mostly have style complaints.

And finally, as @sarbes is the is the expert on this topic, I'll let him make the final decision.

File: xbmc/guilib/AspectRatio.h
Comment (garbear) [reply xbmc/guilib/AspectRatio.h:21]: isn't static not needed for constexpr variables in a header?

Comment (neo1973) [reply xbmc/guilib/AspectRatio.h:21]: Ah yes, you are correct. `inline constexpr` would probably be the best if I my understanding is correct but we haven't done that anywhere and our code guidelines say just `constexpr`as well.

File: xbmc/guilib/GUITextureManager.h
Comment (mxalbert1996) [reply xbmc/guilib/GUITextureManager.h]: Done.

File: xbmc/guilib/GUITextureManager.h
Comment (mxalbert1996) [reply xbmc/guilib/GUITextureManager.h]: Done.
Also I noticed that there is already a `CGUITextureManager` class, so I changed this to `CGUITextureCallbackManager`.

File: xbmc/guilib/GUITextureManager.cpp
Comment (mxalbert1996) [reply xbmc/guilib/GUITextureManager.cpp]: I believe `cbegin()` and `cend()` only prevent direct modification of the current element in iteration, but will not work properly if elements are removed externally during the iteration (which actually happens in `texture->OnWindowResize()`), so we still need a temporary `vector`.

File: xbmc/guilib/FFmpegImage.cpp
Comment (mxalbert1996) [reply xbmc/guilib/FFmpegImage.cpp:490]: Done.

File: xbmc/guilib/AspectRatio.h
Comment (mxalbert1996) [reply xbmc/guilib/AspectRatio.h]: Done.

File: xbmc/guilib/AspectRatio.h
Comment (mxalbert1996) [reply xbmc/guilib/AspectRatio.h]: Made them `inline constexpr`.

File: xbmc/guilib/GUITextureManager.cpp
Comment (neo1973) [reply xbmc/guilib/GUITextureManager.cpp]: Ah, OK. Then please add a short comment why the temporary copy is necessary here.

Comment (mxalbert1996) [reply xbmc/guilib/GUITextureManager.cpp]: Done.

File: xbmc/guilib/FFmpegImage.cpp
Comment (neo1973) [reply xbmc/guilib/FFmpegImage.cpp:490]: Nope, the `struct` is still there :wink:

Comment (mxalbert1996) [reply xbmc/guilib/FFmpegImage.cpp]: Ah, my bad. Fixed.

**Review by neo1973 — APPROVED** [review]
Thanks @mxalbert1996, looks good to me now! But as said I'll leave the decisions to merge to @sarbes :slightly_smiling_face:

**Comment by Bindou01** [comment]
Hi all ! I've been waiting a long time for the implementation of dynamically adjusting artwork resolution based on view type and zoom level. So I'm glad to see this topic discussed here, as it would be a real step forward, especially for low-power devices.

Such a setting would not only optimize CPU usage, but also improve visual quality across different view types.

For now, I've worked around this by forcing Kodi to use a higher resolution for covers. In my experience, 960x1440p is optimal for large views (media info) in most 4K skins; this is double Kodi's base resolution (480x720p). This eliminates blurring in the info boxes displaying large covers, but causes aliasing in widgets or wall views, where covers appear small.

So I wonder why the current downscaling causes this aliasing?

Wouldn't it be easier to download two versions of the artwork (480x720p and 960x1440p) and create a display standard for all skins?

Finally, I don't think it is necessary to download the artwork in its native definition, because most current covers are in 2000x3000 pixels, a size more suited to 8K interfaces than to 4K ones. Limiting the maximum definition via a setting in Kodi would allow for better disk space management and optimized performance.

Prefer 2000X3000 for 8K screens
Prefer 1000X1500 for 4K screens

**Comment by mxalbert1996** [comment]
@Bindou01 
> What is the current downscaling method that is responsible for this aliasing in downscaling? Is this corrected with the changes made here?

Currently Kodi doesn't do any scaling itself for UI images, and just let the GPU scale them, which is the cause of the aliasing.
This PR makes Kodi scale the image to its actual display size before uploading it to the GPU.

> Will the illustrations be downloaded in their native definition with your method? If so, will it be possible to limit the maximum size?

This PR doesn't change artwork downloading and caching.

**Comment by Bindou01** [comment]
> @Bindou01
> 
> > What is the current downscaling method that is responsible for this aliasing in downscaling? Is this corrected with the changes made here?
> 
> Currently Kodi doesn't do any scaling itself for UI images, and just let the GPU scale them, which is the cause of the aliasing. This PR makes Kodi scale the image to its actual display size before uploading it to the GPU.
> 
> > Will the illustrations be downloaded in their native definition with your method? If so, will it be possible to limit the maximum size?
> 
> This PR doesn't change artwork downloading and caching.

Thank you very much for this popularization mxalbert1996.
So if I understand correctly, it is only the scaling algorithm that is discussed here. Which is a great step forward. When could this be integrated into Kodi for the general public ?

However, if I understand correctly, it will still be necessary to go through the advancedsettings.xml configuration file to select the maximum definition of the illustrations. Do we have to create a new request to integrate this directly into the Kodi interface ? If so, where can we make this request ?

**Comment by mxalbert1996** [comment]
> When could this be integrated into Kodi for the general public ?

That's up to the maintainers.

> Do we have to create a new request to integrate this directly into the Kodi interface ? If so, where can we make this request ?

You can, and as said in #25866,
> Feature requests go to the forum please.

**Comment by Bindou01** [comment]
> > When could this be integrated into Kodi for the general public ?
> 
> That's up to the maintainers.
> 
> > Do we have to create a new request to integrate this directly into the Kodi interface ? If so, where can we make this request ?
> 
> You can, and as said in #25866,
> 
> > Feature requests go to the forum please.

Ok, done !
If you are interested in supporting the request :
https://forum.kodi.tv/showthread.php?tid=379241&pid=3213524#pid3213524

Or if you think this is wrong, feel free to let me know.

Thanks !

**Comment by sarbes** [comment]
Sorry for my absence, but I was a bit busy in the last few weeks. 

From a first glance, I'm not so sure that the callback manager is a good idea. The UI elements should decide themself if it makes sense to discard their current texture(s) in order to reload them. There might be textures which can't be downscaled, or where it makes no sense to scale them.

**Comment by Bindou01** [comment]
> Sorry for my absence, but I was a bit busy in the last few weeks.
> 
> From a first glance, I'm not so sure that the callback manager is a good idea. The UI elements should decide themself if it makes sense to discard their current texture(s) in order to reload them. There might be textures which can't be downscaled, or where it makes no sense to scale them.

Would it be possible to provide an option in Kodi settings to reset the artwork to the desired resolution? This option could be accompanied by a clear warning, such as: “Deleting all artwork: irreversible operation”, so that the user is well informed. This would give users an additional choice and make the process easier, as many do not know how to manually adjust the definition of the artwork for 4K screens. Currently, many users end up with blurry artwork, poorly adapted to the resolution of their screen.

Otherwise, in an ideal scenario, resetting the standard resolution of the artwork in Kodi could also be a solution. By adopting, for example, a definition of 960x1440px posters more adapted to common screens in 2024 (4K), this would allow a natural update of the artwork as the user replaces it, and this, without additional configuration.

**Comment by neo1973** [comment]
@Bindou01: That's not at all what the discussion here is about and not in scope for this PR.

**Comment by Bindou01** [comment]
> @Bindou01: That's not at all what the discussion here is about and not in scope for this PR.
Sorry I read too quickly, I wrongly thought Sarbes was answering me. haha

**Comment by mxalbert1996** [comment]
@sarbes 
> The UI elements should decide themself if it makes sense to discard their current texture(s) in order to reload them. There might be textures which can't be downscaled, or where it makes no sense to scale them.

We are only scaling large textures (artwork or photos), which I assume should always be scaled for better quality, shouldn't they?

If we really need to make the reload optional, what do you think of adding something like `CGUITexture::SetShouldReloadOnResize()`? (Since the motivation of adding the callback manager is that triggering the reload from the holders of `CGUITexture` would require too much code change.)

**Comment by Bindou01** [comment]
Why is it blocked, I mean, what changes are being requested?

**Review by tarekk22-afk — APPROVED** [review]

**Review by tarekk22-afk — APPROVED** [review]

**Review by tarekk22-afk — APPROVED** [review]

**Comment by sarbes** [comment]
Sorry for the lack of communication, but I'm short on time these days. I'll trust @neo1973 with the review, as I don't have the capacity to do so.

> We are only scaling large textures (artwork or photos), which I assume should always be scaled for better quality, shouldn't they?

The texture path is also used for skin backdrops and alike. In Estuary, we do have smaller (640x360) backdrops which are scaled to screen resolution. Those would be unloaded each time. But that's not the end of the world

> If we really need to make the reload optional, what do you think of adding something like `CGUITexture::SetShouldReloadOnResize()`? (Since the motivation of adding the callback manager is that triggering the reload from the holders of `CGUITexture` would require too much code change and is more error-prone.)

The logic could be refined later. 

Thank you for your contribution. Are you perhaps interested in improving other shortcomings of our UI system?

**Comment by neo1973** [comment]
Jenkins build this please

**Comment by neo1973** [comment]
Thanks for for your contribution and taking the time to include the requested changes @mxalbert1996.

**Comment by mxalbert1996** [comment]
@sarbes 
> The texture path is also used for skin backdrops and alike.

Oh, I thought those are loaded by `CGUITextureManager` instead of `CGUILargeTextureManager`.

> Are you perhaps interested in improving other shortcomings of our UI system?

Sure!

**Comment by Bindou01** [comment]
Hello and Merry Christmas to all. :)

I see that this feature is now in "merged" status. I'm not really sure what the usual process is. Does this mean that this will necessarily appear in the next Kodi update?

**Comment by the-black-eagle** [comment]
It means it is present in any v22 development builds, available on our download page -> https://kodi.tv/download/  under your chosen platform.

It will not be in any v21 builds.

**Comment by Bindou01** [comment]
Thank you very much. So I was able to compare Kodi 21 with Kodi 22 (Windows), however, I expected there to be a gain in terms of performance thanks to the dynamic display of covers. But I observe identical performances in all respects, even in wall view (small cover). However, the definition of my covers is 1100 x 1640px since advancedsettings, and this has a heavy impact on the framerate, compared to the initial definition. So, I thought that this update would have had positive effects on this point. Is the functionality enabled natively?

**Comment by mxalbert1996** [comment]
@Bindou01 
The goal of this PR is to improve image rendering quality. Memory usage should also be reduced as a side effect. Performance improvement is not the goal.

**Comment by Bindou01** [comment]
I understand that this is not the main goal, but it seemed to me that it should have been a side effect, especially when there are a large number of small covers to display? 🤔 (Regarding the display quality, I don't see any difference either. It's absolutely identical when I compare screenshots between Kodi version 21 and 22). It's like this feature doesn't work.

**Comment by neo1973** [comment]
It is clearly working. 21.1 at the top, master below:
Bildschirmfoto_20241226_135100
It is also noticeably faster on my RPi3B. On my dev machine it doesn't make a difference as it produces 60 fps easily in both versions.

**Comment by chucky88** [comment]
Hi, can it potentially be backported to 21 (even if not officially) or is there a technical obstacle (missing critical dependencies, newer toolchain required, etc)? Thanks!

**Comment by neo1973** [comment]
No such obstacles but expect manual conflict resolving. Definitely doable. But as you guessed, there won't be an official backport to 21 as this is a new feature :slightly_smiling_face:

**Comment by Bindou01** [comment]
Thanks for the information, @neo1973.

After further analysis, I do indeed see a clear improvement in image quality. This improvement is particularly evident on illustrations with strong lines and contrasts.

The aliasing induced by the old method is greatly reduced, resulting in a smoother image, free of artifacts. However, this could result in a slight lack of sharpness in some cases. Regardless, this update is overall excellent. Many thanks to those who worked on this improvement!

Kodi 21 (0)
Kodi 22 (0)

Kodi 21 (1)
Kodi 22 (1)

Kodi 21 (2)
Kodi 22 (2)

I also noticed a gain in memory load, but I have not yet managed to identify a significant impact on graphics load. I understand that this is not the main objective, but I expected a substantial beneficial effect in this area as well. However, I will refine my analysis protocol to confirm this point.

Finally, one last question: could this display method be applied to smaller textures to benefit skin creators as well? I know some skinners continue to develop their skins in 1080p for performance and rendering reasons.

**Comment by Bindou01** [comment]
Finally, one last question: could this display method be applied to smaller textures to benefit skin creators as well? I know some skinners continue to develop their skins in 1080p for performance and rendering reasons.

**Comment by Hitcher** [comment]
> Finally, one last question: could this display method be applied to smaller textures to benefit skin creators as well? I know some skinners continue to develop their skins in 1080p for performance and rendering reasons.

The skin code is converted to the resolution used before any texture scaling happens.

**Pull request merged**
